src/itouch.cpp

```diff
@@ -0,0 +1,399 @@
+{
+    if (_vm.count("replica") && _vm.count("resource")) {
+        std::cerr << "Error: -n and -R cannot be used together.\n";
+        return false;
+    }
+
+    if (_vm.count("reference") && _vm.count("seconds-since-epoch")) {
+        std::cerr << "Error: -r and -s cannot be used together.\n";
+        return false;
+    }
+
+    if (_vm.count("logical_path") == 0) {
+        std::cerr << "Error: Missing logical path.\n";
+        return false;
+    }
+
+    return true;
+}
+
+auto canonical(const std::string& _path, const rodsEnv& _env) -> std::optional<std::string>
+{
+    rodsPath_t input{};
+    rstrcpy(input.inPath, _path.data(), MAX_NAME_LEN);
+
+    if (parseRodsPath(&input, const_cast<rodsEnv*>(&_env)) != 0) {
+        return std::nullopt;
+    }
+
+    auto* escaped_path = escape_path(input.outPath);
+    std::optional<std::string> p = escaped_path;
+    std::free(escaped_path);
+
+    return p;
+}
+
+auto get_time(ix::client_connection& _conn,
+              const rodsEnv& _env,
+              const po::variables_map& _vm) -> fs::object_time_type
+{
+    if (_vm.count("reference")) {
+        std::string ref_path;
+
+        if (auto v = canonical(_vm["reference"].as<std::string>(), const_cast<rodsEnv&>(_env)); v) {
+            ref_path = *v;
+        }
+        else {
+            throw std::runtime_error{"Failed to convert reference path to absolute path."};
+        }
+
+        return fs::client::last_write_time(_conn, ref_path);
+    }
+
+    using duration_type = fs::object_time_type::duration;
+
+    if (_vm.count("seconds-since-epoch")) {
+        return fs::object_time_type{duration_type{_vm["seconds-since-epoch"].as<rep_type>()}};
+    }
+
+    return std::chrono::time_point_cast<duration_type>(fs::object_time_type::clock::now());
+}
+
+auto create_data_object_if_necessary(ix::client_connection& _conn,
+                                     const rodsEnv& _env,
+                                     const po::variables_map& _vm,
+                                     const fs::path& _path) -> void
+{
+    if (fs::client::exists(_conn, _path)) {
+        return;
+    }
+
+    // Return immediately if the object does not exist and the user specified to not
+    // create any new data objects.
+    if (_vm.count("no-create")) {
+        throw std::runtime_error{"Cannot update modification time (object does not exist)."};
+    }
+
+    if (_vm.count("replica")) {
+        throw std::runtime_error{"Replica numbers cannot be used when creating new data objects."};
+    }
+
+    // Create a new data object if the path does not end with a path separator.
+    io::client::default_transport tp{_conn};
+
+    // Create the data object at the leaf resource if the user provided one.
+    if (_vm.count("resource")) {
+        io::odstream{tp, _path, io::leaf_resource_name{_vm["resource"].as<std::string>()}};
+        return;
+    }
+
+    if (std::strlen(_env.rodsDefResource) == 0) {
```

Comment:
This needs to be revisited. There is likely a much better/safer way to test for empty string.

Reply:
can't the server determine where to create the data object in the absence of a default coming from the client?

Reply:
That's the thing, there is no server side code to check this. `itouch` is implememted entirely on the client side. Which does mean the microservice version has to be implemented from the ground up.

We may have to create a new API plugin and move all of this into it. Then the clients can just invoke an API endpoint and everything behaves the same.